### ISSUE TITLE
TopoLM.py

### ISSUE THREAD

**Issue by shandingwangyue** [shandingwangyue/TopoLM#2]
"""
TopoLM: Topological Domain Attention - Proof of Concept
-------------------------------------------------------
This script demonstrates the core mechanism of the TopoLM architecture:
Spatial-Indexed Attention via Topological Domain Embeddings.

It simulates a scenario where an immutable Macro-Domain (e.g., Physics/Thermodynamics) 
filters incoming micro-entity vectors. Conceptually illegal tokens (e.g., magic, 
perpetual motion) are mathematically rejected (masked) before the attention calculation.
"""

import torch
import torch.nn as nn
import torch.nn.functional as F

class TopologicalDomainAttention(nn.Module):
    """
    A modified Multi-Head Attention layer that incorporates Topological Domain constraints.
    Instead of calculating O(N^2) attention blindly, it calculates the inclusion probability
    of each token within the active Macro-Domain and masks out structurally illegal tokens.
    """
    def __init__(self, d_model, num_heads):
        super().__init__()
        self.d_model = d_model
        self.num_heads = num_heads
        
        # Standard Q, K, V projections
        self.qkv_proj = nn.Linear(d_model, 3 * d_model)
        self.out_proj = nn.Linear(d_model, d_model)

    def forward(self, x, domain_mu, domain_sigma, threshold=0.05):
        """
        Args:
            x: Input token vectors of shape [batch_size, seq_len, d_model]
            domain_mu: Semantic center of the active Macro-Domain [d_model]
            domain_sigma: Variance/Volume of the active Macro-Domain [d_model]
            threshold: Probability density threshold for topological inclusion
        """
        batch_size, seq_len, _ = x.size()

        # ---------------------------------------------------------
        # STEP 1: Calculate Topological Inclusion (Probability Density)
        # Using a simplified squared Mahalanobis distance for multivariate Gaussian
        # ---------------------------------------------------------
        diff = x - domain_mu
        mahalanobis_sq = (diff ** 2) / (domain_sigma + 1e-9) # Add epsilon for numerical stability
        
        # Calculate scaled inclusion probability [0, 1]
        inclusion_prob = torch.exp(-0.5 * mahalanobis_sq.mean(dim=-1)) 

        # ---------------------------------------------------------
        # STEP 2: Generate Spatial Mask
        # 1 if token strictly belongs to the active domain, 0 otherwise
        # ---------------------------------------------------------
        domain_mask = (inclusion_prob > threshold).float()
        
        # Expand mask for multi-head attention matrix [batch, heads, seq_len, seq_len]
        # This prevents any token from attending to out-of-domain tokens
        attn_mask = domain_mask.unsqueeze(1).unsqueeze(2) 

        # ---------------------------------------------------------
        # STEP 3: Standard QKV Projection & Attention Calculation
        # ---------------------------------------------------------
        qkv = self.qkv_proj(x).chunk(3, dim=-1)
        q, k, v = map(lambda t: t.view(batch_size, seq_len, self.num_heads, -1).transpose(1, 2), qkv)

        # Raw attention scores
        scores = torch.matmul(q, k.transpose(-2, -1)) / (q.size(-1) ** 0.5)

        # APPLY TOPOLOGICAL DOMAIN MASK: Set invalid relationships to -infinity
        scores = scores.masked_fill(attn_mask == 0, float('-inf'))

        # Only structurally legal tokens proceed through softmax
        attn_weights = F.softmax(scores, dim=-1)
        out = torch.matmul(attn_weights, v)

        # Reshape and project output
        out = out.transpose(1, 2).contiguous().view(batch_size, seq_len, self.d_model)
        
        return self.out_proj(out), inclusion_prob

# =====================================================================
# Experimental Simulation: Defending Physics Axioms vs. Hallucinations
# =====================================================================
if __name__ == "__main__":
    # Set random seed for reproducibility
    torch.manual_seed(42)
    
    # Model parameters
    d_model = 256
    num_heads = 8
    
    print("\n[TopoLM Simulation] Initializing Topological Domain Attention...")
    topo_attn = TopologicalDomainAttention(d_model, num_heads)
    
    # -----------------------------------------------------------------
    # Define Immutable Macro-Domain Prior: D_physics (Thermodynamics)
    # The center is set to 0, with strict (small) variance defining its boundaries.
    # -----------------------------------------------------------------
    mu_physics = torch.zeros(d_model)
    sigma_physics = torch.ones(d_model) * 0.5 
    
    # -----------------------------------------------------------------
    # Simulate Input Tokens (Micro Entities)
    # -----------------------------------------------------------------
    # Token 0 ("engine") and Token 1 ("efficiency"): 
    # Reality/Physics based concepts, geometrically close to mu_physics
    valid_tokens = torch.randn(1, 2, d_model) * 0.4
    
    # Token 2 ("magic_qi") and Token 3 ("perpetual_motion"): 
    # Fiction/Impossible concepts, geometrically distant from mu_physics
    invalid_tokens = torch.randn(1, 2, d_model) * 3.0 + 8.0 
    
    # Combine into a single sequence
    tokens = torch.cat([valid_tokens, invalid_tokens], dim=1)
    token_names = ["engine", "efficiency", "magic_qi", "perpetual_motion"]
    
    print(f"\n[Context Activated] Layer 1 Domain: Physics/Thermodynamics")
    print(f"[Input Sequence] {token_names}\n")
    
    # -----------------------------------------------------------------
    # Forward Pass through the active Physics Domain
    # -----------------------------------------------------------------
    # Using a strict threshold to enforce rigorous physical laws
    output, probs = topo_attn(tokens, mu_physics, sigma_physics, threshold=0.01)
    
    # Display results
    print("--- Topological Inclusion Probabilities & Masking Results ---")
    for i, p in enumerate(probs[0]):
        status = "ACCEPTED (Proceeds to Self-Attention)" if p > 0.01 else "REJECTED (Attention Masked)"
        print(f"Token {i} [{token_names[i]:<16}]: Prob = {p.item():.4f} -> {status}")
        
    print("\n[Conclusion] The model mathematically refuses to calculate relationships involving 'perpetual_motion' within the Physics domain, preventing hallucination via structural friction.\n")
